mathesar_ui/src/component-library/multi-select/MultiSelect.svelte

```diff
@@ -0,0 +1,137 @@
+  import { getGloballyUniqueId } from '@mathesar-component-library-dir/common/utils/domUtils';
+  import type { LabelGetter } from '@mathesar-component-library-dir/common/utils/formatUtils';
+  import { getLabel as defaultGetLabel } from '@mathesar-component-library-dir/common/utils/formatUtils';
+  import { AttachableDropdown } from '@mathesar-component-library-dir/dropdown';
+  import Icon from '@mathesar-component-library-dir/icon/Icon.svelte';
+  import {
+    ListBox,
+    ListBoxOptions,
+  } from '@mathesar-component-library-dir/list-box';
+  import StringOrComponent from '@mathesar-component-library-dir/string-or-component/StringOrComponent.svelte';
+
+  type Option = $$Generic;
+
+  const dispatch = createEventDispatcher<{ change: Option[] }>();
+
+  export let id = getGloballyUniqueId();
+
+  export let disabled = false;
+
+  /**
+   * Specifies the key on which the options label is stored.
+   */
+  export let labelKey = 'label';
+
+  export let getLabel: LabelGetter<Option> = (o: Option) =>
+    defaultGetLabel(o, labelKey);
+
+  /**
+   * List of options to select from.
+   */
+  export let options: Option[] = [];
+
+  export let values: Option[] = [];
+
+  /**
+   * Classes to apply to the content (each of the options).
+   */
+  export let contentClass = '';
+
+  /**
+   * Classes to apply to the trigger button (the dropdown button).
+   */
+  export let triggerClass = '';
+
+  /**
+   * The ARIA label for this select component.
+   */
+  export let ariaLabel: string | undefined = undefined;
+
+  /**
+   * By default, options will be compared by equality. If you're using objects as
+   * options, you can supply a custom function here to compare them.
+   *
+   * For example:
+   *
+   * ```ts
+   * valuesAreEqual={(a, b) => a.id === b.id}
+   * ```
+   */
+  export let valuesAreEqual: (
+    optionToCompare: Option | undefined,
+    selectedOption: Option | undefined,
+  ) => boolean = (a, b) => a === b;
+
+  let trigger: HTMLElement;
+
+  function setValues(newValues: Option[]) {
+    values = newValues;
+    dispatch('change', values);
+  }
+
+  function removeValue(value: Option) {
+    setValues(values.filter((v) => !valuesAreEqual(v, value)));
+  }
+
+  function handleOptionsChange(newOptions: Option[]) {
+    setValues(
+      values.filter((v) => newOptions.some((o) => valuesAreEqual(o, v))),
+    );
+  }
+  $: handleOptionsChange(options);
+</script>
+
+<BaseInput {...$$restProps} {id} {disabled} />
+
+<ListBox
+  {options}
+  bind:value={values}
+  on:change
+  {labelKey}
+  {getLabel}
+  checkEquality={valuesAreEqual}
+  let:api
+  let:isOpen
+>
+  <span
+    class="multi-select-trigger {triggerClass}"
+    bind:this={trigger}
+    on:click={() => api.toggle()}
+    aria-label={ariaLabel}
+  >
+    <span class="selected-values">
+      {#each values as value}
+        <span class="selected-value">
+          <span class="label">
+            <StringOrComponent arg={getLabel(value)} />
```

Comment:
Ok, I'm open to this. I'd like to understand more though.

> Svelte cannot perform any optimizations on it.

Can you point me towards the resources which lead you to the conclusion that usage of `svelte:component` has downsides? I'd like to better understand these downsides.

> it can avoided here

I'd like to better understand the pattern you want to employ. Perhaps you could open a PR, making the changes yourself?

> I've updated `ListBoxOptions` and `Select` components

I'm confused by this because I see that `ListBoxOptions` and `Select` both still contain `StringOrComponent`.